### PR TITLE
[Messenger] Added UniqueStamp to help ensure ExactlyOnce semantics

### PR DESCRIPTION
**Pull request by ragboyjr**
The unique stamp can be used to declare that a given message
should be unique inside of a queue awaiting processing. This
stamp should allow ensuring that certain messages don't get
double processed if they are already awaiting in the queue.

Example:

- You have a queue for syncing product updates to an external
  service
- On every product save, you enqueue an UpdatedProductMessage
- The unique stamp would make sure that saving the same product
  multiple times will result in only one message in the queue
  preventing duplicated messages.

Signed-off-by: RJ Garcia <ragboyjr@icloud.com>

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | TKTK

### CODE REVIEW

File: src/Symfony/Component/Messenger/Stamp/UniqueStamp.php
Comment (Koc) [inline src/Symfony/Component/Messenger/Stamp/UniqueStamp.php]: What sense add this stamp without providing of id?

Comment (ragboyjr) [reply src/Symfony/Component/Messenger/Stamp/UniqueStamp.php]: The transport could generate a unique id from like a hash of the message body. SQS for example supports both client supplied ids or allowing sqs to dedup it on their end.

Comment (ragboyjr) [reply src/Symfony/Component/Messenger/Stamp/UniqueStamp.php]: But this is definitely something I should make clear in the documentation.

Comment (OskarStark) [reply src/Symfony/Component/Messenger/Stamp/UniqueStamp.php:25]: Is it needed to enforce string or null? Could we allow int?

File: src/Symfony/Component/Messenger/Tests/Stamp/UniqueStampTest.php
Comment (OskarStark) [inline src/Symfony/Component/Messenger/Tests/Stamp/UniqueStampTest.php:28]: No I’d given

Comment (ragboyjr) [reply src/Symfony/Component/Messenger/Tests/Stamp/UniqueStampTest.php]: oh derp, this is fixed now. Thanks for spotting this!

File: src/Symfony/Component/Messenger/Stamp/UniqueStamp.php
Comment (ragboyjr) [reply src/Symfony/Component/Messenger/Stamp/UniqueStamp.php:25]: Hmm, i'm not sure, SQS uses strings up to 100 chars, and if the id is null, it's expected to be a hash of the serialized body which is also a string, and ints are easily typecast-able to strings.

If I could typehint `?int|string`, then I think it'd be a no brainer, but that's not available till php 8. I definitely prefer getting to use a php type decl vs a docblock, but if the consensus of the core team is that we should allow both, then i'm cool with that.

**Review by Tobion — CHANGES_REQUESTED** [review]
This PR only adds the stamp but it's not integrated in messenger in any way. Just the stamp does not do anything.

We implemented a feature like this ourselves. But it's using a semaphore to actually ensure uniqueness.
So without such integration, this seems not that useful at the moment.
Related to #35780

**Comment by ragboyjr** [comment]
@Tobion Very fair point. Here's an example of a Redis adapter that uses redis lists (not streams) that makes use of the unique stamp: https://github.com/krakphp/symfony-messenger-redis/blob/master/src/Transport/RedisTransport.php#L137

I figured, it'd be something that we'd roll out incrementally to the various adapters after merge, but if you feel like we should do them at more/less the same time, that's fine as well.

**Comment by ragboyjr** [comment]
Also, how exactly are you using a semaphore to ensure uniqueness of a specific message while it's in the queue waiting to be processed? Having a hard time wrapping my head around that.

File: src/Symfony/Component/Messenger/Stamp/UniqueStamp.php
Comment (OskarStark) [reply src/Symfony/Component/Messenger/Stamp/UniqueStamp.php:25]: I am definitely 👍🏻 for a typehint but instead of just having a docblock you could check the type inside the method

**Comment by fabpot** [comment]
@Tobion Can you help @ragboyjr figues out how to improve this PR?

File: src/Symfony/Component/Messenger/Stamp/UniqueStamp.php
Comment (jderusse) [inline src/Symfony/Component/Messenger/Stamp/UniqueStamp.php:21]: AwsSQS transport already provides a `AmazonSqsFifoStamp` to achieve this behavior.
Making this a final class, prevent using this "generic" behavior with AwsSQS.

**Comment by sroze** [comment]
@ragboyjr would you be able to describe how you intend existing transports to use this stamp?

**Comment by ragboyjr** [comment]
Well, for redis, I've already done something similar for a list based implementation here: https://github.com/krakphp/symfony-messenger-redis/blob/master/src/Transport/RedisTransport.php#L118

Essentially, I'm just using a lua script to ensure that when we push a message with a unique id, we add the unique id to a set, if it already exists in the set, we don't add the message to the queue.
When we pop an item from the queue, we remove it from the set if it contained a unique id.

For SQS, they have the concept of a MessageDeDuplicationId which we could set in this case.

For rabbit mq, if you use the de-duplication plugin, you could just pass the id directly into dedup header.

For the db driver, this would be fairly straightforward to implement by just including a message id column or something similar.

**Comment by weaverryan** [comment]
I like this idea! But I agree that we should have full integration before merging it (not that I know how to do that...)

**Comment by fabpot** [comment]
We've just moved away from `master` as the main branch to use `5.x` instead. Unfortunately, I cannot reopen the PR and change the target branch to `5.x`. Can you open a new PR referencing this one to not loose the discussion? Thank you for your understanding and for your help.

**Comment by ragboyjr** [comment]
Thanks all, will open a new PR and start working integrating this feature with other transports

**Pull request closed without merging**
